pkg/auth/authmap_gc.go

```diff
@@ -39,7 +39,10 @@ type authMapGarbageCollector struct {
 	ciliumIdentitiesSynced     bool
 	ciliumIdentitiesDeleted    map[identity.NumericIdentity]struct{}
 
-	endpointRepo endpointRepository
+	endpointRepo         endpointRepository
+	endpointsCache       map[uint16]*endpoint.Endpoint
```

Comment:
We could improve this by only keeping the relevant information in the cache.

Instead of the whole Endpoint we should only keep its SecurityIdentity as value of this map. WDYT?

This way we preprocess the relevant data in the event handler instead of the GC execution itself.

Reply:
I'm not sure whether the SecurityIdentity of an EP can change during it's lifetime. If so, that could break stuff if we only keep the ID instead of the full EP reference, I think?

Reply:
Good point. Would be great to know anyway - definitely worth a comment if this is the case. Maybe @christarazi can clarify this?

This is a good additional reason to keep the data kept in the GarbageCollector as little as possible - to get rid of these potential side-effects.

Reply:
this would need some kind of "reference counting" and i am indeed not sure on security IDs changing, they can as we deal with pointers. In the first attempt of this i did something similar and got out of sync issues because of it, though i deleted all that code so couldn't tell exactly what i did.

Reply:
> this would need some kind of "reference counting"

_Assuming_ that the ep -> security id is stable: Why should it be necessary to count the references?

In the cleanup, it's only necessary to to get all security ids that are in use -> collecting & merging all values from the map `map[uint16]*identity.Identity` (ep -> SID)

Reply:
Endpoint security identities can certainly change. The most common way is if the user modifies the labels for the underlying pod.

Reply:
Thanks for the info! Interesting that there's no "update" function on the subscriber interface.